{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e117ac4e-89ed-40c1-b5a9-8caafaa55632",
   "metadata": {},
   "source": [
    "Rnaseq analysis, Report generated by Mathu Malar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d087acc-2202-4b3f-af73-21d239a0ea88",
   "metadata": {},
   "outputs": [],
   "source": [
    "step1: Data Qc using Fastqc and Multiqc [ Note: Snakemake was used and script is attached below]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7085c96c-bb7c-463f-b0df-1fa3d8592c2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLES = [\"0H_Infected_Rep1\", \"0H_Infected_Rep2\", \"0H_Mock_Rep1\",\"0H_Mock_Rep2\",\"48H_Infected_Rep1\",\"48H_Infected_Rep2\",\"48H_Mock_Rep1\",\"48H_Mock_Rep2\",\"72H_Infected_Rep1\",\"72H_Infected_Rep2\",\"72H_Mock_Rep1\",\"72H_Mock_Rep2\"]\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand(\"/home/malarm/RNAseq/test/output/{sample}_R1_fastqc.zip\", sample=SAMPLES),\n",
    "        expand(\"/home/malarm/RNAseq/test/output/{sample}_R2_fastqc.zip\", sample=SAMPLES),\n",
    "        \"/home/malarm/RNAseq/test/output/multiqc_report.html\"\n",
    "\n",
    "\n",
    "\n",
    "rule fastqc:\n",
    "    input:\n",
    "        \"/home/malarm/RNAseq/{sample}_R1.fastq.gz\",\n",
    "        \"/home/malarm/RNAseq/{sample}_R2.fastq.gz\"\n",
    "    output:\n",
    "        \"/home/malarm/RNAseq/test/output/{sample}_R1_fastqc.zip\",\n",
    "        \"/home/malarm/RNAseq/test/output/{sample}_R2_fastqc.zip\"\n",
    "    shell:\n",
    "        \"fastqc {input} --outdir output\"\n",
    "\n",
    "\n",
    "rule multiqc:\n",
    "    input:\n",
    "        expand(\"/home/malarm/RNAseq/test/output/{sample}_R1_fastqc.zip\", sample=SAMPLES),\n",
    "        expand(\"/home/malarm/RNAseq/test/output/{sample}_R2_fastqc.zip\", sample=SAMPLES)\n",
    "    output:\n",
    "        \"/home/malarm/RNAseq/test/output/multiqc_report.html\"\n",
    "    shell:\n",
    "        \"multiqc -o output --filename multiqc_report {input}\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fef87406-8fcf-4286-bef7-f49d38e1b36a",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "168c2620-5fba-41e9-99cd-0a6e06d1a772",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<a href=\"https://drive.google.com/file/d/1Rx5JeE9olQsdejznDGf4mCbujOMB42O9/view?usp=drive_link\">View HTML Document</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "\n",
    "display(HTML('<a href=\"https://drive.google.com/file/d/1Rx5JeE9olQsdejznDGf4mCbujOMB42O9/view?usp=drive_link\">View HTML Document</a>'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96c46f6c-d1ed-447e-98f1-c1b94f2c3c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Step2: Trimming using fastp [ Note: shell command attached here]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae7c52c7-fee6-4722-a9a3-1a52cea4e236",
   "metadata": {},
   "outputs": [],
   "source": [
    "for first in *_R1.fastq.gz; do\n",
    "    second=$(echo \"${first}\" | sed 's/_R1/_R2/g')\n",
    "    output1=\"${first%_R1.fastq.gz}_R1trimmed.fastq.gz\"\n",
    "    output2=\"${second%_R2.fastq.gz}_R2trimmed.fastq.gz\"\n",
    "    ../fastp -i \"${first}\" -I \"${second}\" -o \"${output1}\" -O \"${output2}\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a6afaf77-36ef-42be-a8e8-aa3eb6b42225",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<a href=\"https://drive.google.com/file/d/1nPp5-VLXlc8bfb3Mb9o5uqTsnbSMZ5WT/view?usp=drive_link\">View HTML Document</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "\n",
    "display(HTML('<a href=\"https://drive.google.com/file/d/1nPp5-VLXlc8bfb3Mb9o5uqTsnbSMZ5WT/view?usp=drive_link\">View HTML Document</a>'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a57226de-4d1b-4849-8278-9c3d867b5852",
   "metadata": {},
   "outputs": [],
   "source": [
    "Step3: Trimming with fastp worsens length distribution of the fastq sequences so need to give it a thinking does it need trimming? reports are attached below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e51023d4-4879-476e-afa9-a0c5bcdb1e15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<a href=\"https://drive.google.com/file/d/19uEK_Sjsj9cu5448Eu8RsN6nm50SZ0bP/view?usp=drive_link\">View HTML Document</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "\n",
    "display(HTML('<a href=\"https://drive.google.com/file/d/19uEK_Sjsj9cu5448Eu8RsN6nm50SZ0bP/view?usp=drive_link\">View HTML Document</a>'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da5a7f88-9b41-4a87-9b11-5421d5bcafc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "checking overrepresented sequences with blast figured out that first one is from A. thaliana genome didn't remove it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b73c4013-e259-4992-9963-675b69d25624",
   "metadata": {},
   "outputs": [],
   "source": [
    "TTTCATATGAGATCGGACCAGCAGGCCATAGTGGTTCTGAGCGACGCTAA\n",
    "TTCGTCAATAGCTGAAGCTTAGTGGCTTTTCTTGGAAGCAGGGTATCACT\n",
    "CGGTCGTCAATAGCTGAAGCTTAGTGGCTTTTCTTGGAAGCAGGGTATCA"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a62dc468-cd9d-4806-8f57-77d028f17ed9",
   "metadata": {},
   "source": [
    "running fastp with an option of providing the contaminated sequence like this and trimming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c4d8888-164f-483f-9856-c92ea8c2da5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for first in *_R1.fastq.gz; do\n",
    "    second=$(echo \"${first}\" | sed 's/_R1/_R2/g')\n",
    "    output1=\"${first%_R1.fastq.gz}_R1trimmed.fastq.gz\"\n",
    "    output2=\"${second%_R2.fastq.gz}_R2trimmed.fastq.gz\"\n",
    "    ../fastp -i \"${first}\" -I \"${second}\" -o \"${output1}\" --adapter_fasta /home/malarm/RNAseq/overrep_seq.fa  -O \"${output2}\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "29f03c64-4f16-4018-9906-51a848f18382",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<https://drive.google.com/file/d/1HSgx7XGaGqbrSpF5oaqMdX70Q7xxJF9_/view?usp=sharing>View HTML Document</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "\n",
    "display(HTML('<https://drive.google.com/file/d/1HSgx7XGaGqbrSpF5oaqMdX70Q7xxJF9_/view?usp=sharing>View HTML Document</a>'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3d5a883-8fb9-4a5e-b595-668b5e29bce6",
   "metadata": {},
   "source": [
    "Step4: Mapping the trimmed fastq file with Arab reference genome using STAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c32f9810-0f36-4f5a-9f88-f8c1f1ed4238",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/home/malarm/STAR-2.7.11b/bin/Linux_x86_64/STAR --runMode genomeGenerate --genomeSAindexNbases 12 --genomeDir /home/malarm/RNAseq/trimmed/test --genomeFastaFiles /home/malarm/RNAseq/trimmed/test/TAIR10_Arabidopsis_Genome.fas\n",
      "\tSTAR version: 2.7.11b   compiled: 2024-01-25T16:12:02-05:00 :/home/dobin/data/STAR/STARcode/STAR.master/source\n",
      "Feb 14 23:09:57 ..... started STAR run\n",
      "Feb 14 23:09:57 ... starting to generate Genome files\n",
      "Feb 14 23:09:59 ... starting to sort Suffix Array. This may take a long time...\n",
      "Feb 14 23:09:59 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Process is interrupted.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Set the path to the STAR installation directory\n",
    "STAR_HOME=\"/home/malarm/STAR-2.7.11b/bin/Linux_x86_64\"\n",
    "\n",
    "# Set the path to the STAR reference genome index\n",
    "REFERENCE_GENOME=\"/home/malarm/RNAseq/trimmed/test\"\n",
    "\n",
    "# Set the input and output directories\n",
    "INPUT_DIR=\"/home/malarm/RNAseq/trimmed/test\"\n",
    "OUTPUT_DIR=\"/home/malarm/RNAseq/trimmed/test\"\n",
    "\n",
    "# Build STAR index\n",
    "$STAR_HOME/STAR --runMode genomeGenerate --genomeSAindexNbases 12 --genomeDir $REFERENCE_GENOME --genomeFastaFiles $REFERENCE_GENOME/TAIR10_Arabidopsis_Genome.fas\n",
    "\n",
    "# Loop over all FASTQ files in the input directory\n",
    "for sample in \"$INPUT_DIR\"/*_R1trimmed.fastq; do\n",
    "    # Extract the sample name without extension\n",
    "    sample_name=$(basename \"$sample\" _R1trimmed.fastq)\n",
    "\n",
    "    # Run STAR on paired-end reads and output SAM file\n",
    "    $STAR_HOME/STAR --genomeDir $REFERENCE_GENOME --readFilesIn \"$INPUT_DIR\"/\"${sample_name}_R1trimmed.fastq\" \"$INPUT_DIR\"/\"${sample_name}_R2trimmed.fastq\" --outFileNamePrefix \"$OUTPUT_DIR\"/\"${sample_name}_aligned\" --outSAMtype BAM SortedByCoordinate\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c286a15-16d6-4703-9f26-e308e9230d1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Step5:counting the number of genes mapped using featurecounts , the output is count file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28fe5453-efe3-4c5b-860d-3decff9eb2a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "featureCounts -t exon -g gene_id -a Araport11_GTF_genes_transposons.current.gtf -o count_out.txt *.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99df29c0-22ef-43a9-acce-a472f51e52d8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
